File: webflux-dialogue/src/main/java/com/study/webflux/rag/domain/memory/port/VectorMemoryPort.java
package com.study.webflux.rag.domain.memory.port;

import java.util.List;

import com.study.webflux.rag.domain.memory.model.Memory;
import com.study.webflux.rag.domain.memory.model.MemoryImportanceUpdateCommand;
import com.study.webflux.rag.domain.memory.model.VectorMemorySearchQuery;
import reactor.core.publisher.Flux;
import reactor.core.publisher.Mono;

public interface VectorMemoryPort {

	Mono<Memory> upsert(Memory memory, List<Float> embedding);

	Flux<Memory> search(VectorMemorySearchQuery query);

	Mono<Void> updateImportance(MemoryImportanceUpdateCommand command);
}

File: app-miyou/src/main/kotlin/com/miyou/app/domain/memory/port/VectorMemoryPort.kt
package com.miyou.app.domain.memory.port

import com.miyou.app.domain.dialogue.model.ConversationSessionId
import com.miyou.app.domain.memory.model.Memory
import com.miyou.app.domain.memory.model.MemoryType
import reactor.core.publisher.Flux
import reactor.core.publisher.Mono
import java.time.Instant

interface VectorMemoryPort {
    fun upsert(
        memory: Memory,
        embedding: List<Float>,
    ): Mono<Memory>

    fun search(
        sessionId: ConversationSessionId,
        queryEmbedding: List<Float>,
        types: List<MemoryType>,
        importanceThreshold: Float,
        topK: Int,
    ): Flux<Memory>

    fun updateImportance(
        memoryId: String,
        newImportance: Float,
        lastAccessedAt: Instant,
        accessCount: Int,
    ): Mono<Void>
}
